src/Sarif/Readers/DeferredDictionaryConverter.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    ///  DeferredDictionaryConverter is a JsonConverter which allows reading IDictionary&lt;string, T&gt;
+    ///  items only at enumeration time, saving the memory cost of keeping every object around.
+    ///  
+    ///  The set of keys and the file position in the JSON stream of the values is pre-loaded and kept around
+    ///  for fast retrieval.
+    /// </summary>
+    /// <typeparam name="T">Type of items in the Dictionary</typeparam>
+    public class DeferredDictionaryConverter<T> : JsonConverter
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(IDictionary<string, T>);
```

Comment:
>objectType [](start = 19, length = 10)

Not sure how important this is but your converter on surface should work for both IDictionary<string, T> and the concrete Dictionary<string, T> types in both read and write cases. you could adjust your CanConvert implementation to reflect this, either by using IsAssignableTo or by adding a second typeof comparison.